### PR TITLE
Update to be able to run on 3.4 with frozen strings

### PR DESCRIPTION
**Pull request by npezza93**

### CODE REVIEW

File: lib/kamal/commands/base.rb [hunk 1]
```diff
@@ -11,13 +11,13 @@ def initialize(config)
...
+      "ssh".then do |cmd|
         if config.ssh.proxy && config.ssh.proxy.is_a?(Net::SSH::Proxy::Jump)
-          cmd << " -J #{config.ssh.proxy.jump_proxies}"
+          cmd = "#{cmd} -J #{config.ssh.proxy.jump_proxies}"
         elsif config.ssh.proxy && config.ssh.proxy.is_a?(Net::SSH::Proxy::Command)
-          cmd << " -o ProxyCommand='#{config.ssh.proxy.command_line_template}'"
+          cmd = "#{cmd} -o ProxyCommand='#{config.ssh.proxy.command_line_template}'"
         end
-        cmd << " -t #{config.ssh.user}@#{host} -p #{config.ssh.port} '#{command.join(" ").gsub("'", "'\\\\''")}'"
+        "#{cmd} -t #{config.ssh.user}@#{host} -p #{config.ssh.port} '#{command.join(" ").gsub("'", "'\\\\''")}'"
```
Comment (jeremy) [inline lib/kamal/commands/base.rb:20]: Would extract an `ssh_proxy_args` method to sidestep this:
```ruby
def run_over_ssh(*command, host:)
  "ssh #{ssh_proxy_args} -t …"
end

def ssh_proxy_args
  case config.ssh.proxy
  when Net::SSH::Proxy::Jump
    "-J #{config.ssh.proxy.jump_proxies}"
  when Net::SSH::Proxy::Command
    "-o ProxyCommand='#{config.ssh.proxy.command_line_template}'"
  end
end
```

File: lib/kamal/commands/base.rb [hunk 2]
```diff
@@ -92,5 +85,13 @@ def grep(*args)
...
         Kamal::Tags.from_config(config, **details)
       end
+
+      def ssh_proxy_args
+        if config.ssh.proxy && config.ssh.proxy.is_a?(Net::SSH::Proxy::Jump)
+          "-J #{config.ssh.proxy.jump_proxies}"
+        elsif config.ssh.proxy && config.ssh.proxy.is_a?(Net::SSH::Proxy::Command)
+          "-o ProxyCommand='#{config.ssh.proxy.command_line_template}'"
+        end
+      end
```
Comment (Sija) [inline lib/kamal/commands/base.rb:95]:
```suggestion
      def ssh_proxy_args
        case config.ssh.proxy
        when Net::SSH::Proxy::Jump
          "-J #{config.ssh.proxy.jump_proxies}"
        when Net::SSH::Proxy::Command
          "-o ProxyCommand='#{config.ssh.proxy.command_line_template}'"
        end
      end
```

File: lib/kamal/commands/base.rb [hunk 3]
```diff
@@ -92,5 +85,13 @@ def grep(*args)
       def tags(**details)
         Kamal::Tags.from_config(config, **details)
       end
+
+      def ssh_proxy_args
+        case config.ssh.proxy
+        when Net::SSH::Proxy::Jump then "-J #{config.ssh.proxy.jump_proxies}"
```
Comment (Sija) [inline lib/kamal/commands/base.rb:91]:
```suggestion
        when Net::SSH::Proxy::Jump
          "-J #{config.ssh.proxy.jump_proxies}"
```

File: lib/kamal/commands/base.rb [hunk 4]
```diff
@@ -11,14 +11,7 @@ def initialize(config)
...
     def run_over_ssh(*command, host:)
-      "ssh".tap do |cmd|
-        if config.ssh.proxy && config.ssh.proxy.is_a?(Net::SSH::Proxy::Jump)
-          cmd << " -J #{config.ssh.proxy.jump_proxies}"
-        elsif config.ssh.proxy && config.ssh.proxy.is_a?(Net::SSH::Proxy::Command)
-          cmd << " -o ProxyCommand='#{config.ssh.proxy.command_line_template}'"
-        end
-        cmd << " -t #{config.ssh.user}@#{host} -p #{config.ssh.port} '#{command.join(" ").gsub("'", "'\\\\''")}'"
-      end
+      "ssh #{ssh_proxy_args} -t #{config.ssh.user}@#{host} -p #{config.ssh.port} '#{command.join(" ").gsub("'", "'\\\\''")}'".squish
```
Comment (jeremy) [inline lib/kamal/commands/base.rb:14]:
```suggestion
      "ssh#{ssh_proxy_args} -t #{config.ssh.user}@#{host} -p #{config.ssh.port} '#{command.join(" ").gsub("'", "'\\\\''")}'".squish
```

File: lib/kamal/commands/base.rb [hunk 5]
```diff
@@ -92,5 +85,14 @@ def grep(*args)
       def tags(**details)
         Kamal::Tags.from_config(config, **details)
       end
+
+      def ssh_proxy_args
+        case config.ssh.proxy
+        when Net::SSH::Proxy::Jump
+          "-J #{config.ssh.proxy.jump_proxies}"
```
Comment (jeremy) [inline lib/kamal/commands/base.rb:92]:
```suggestion
          " -J #{config.ssh.proxy.jump_proxies}"
```

File: lib/kamal/commands/base.rb [hunk 6]
```diff
@@ -92,5 +85,14 @@ def grep(*args)
...
+        when Net::SSH::Proxy::Command
+          "-o ProxyCommand='#{config.ssh.proxy.command_line_template}'"
```
Comment (jeremy) [inline lib/kamal/commands/base.rb:94]:
```suggestion
          " -o ProxyCommand='#{config.ssh.proxy.command_line_template}'"
```

**Review by jeremy — CHANGES_REQUESTED** [review]
Tweak spacing for clean output (and so tests pass)

File: lib/kamal/commands/base.rb [hunk 7]
```diff
@@ -11,7 +11,7 @@ def initialize(config)
     end
 
     def run_over_ssh(*command, host:)
-      "ssh #{ssh_proxy_args} -t #{config.ssh.user}@#{host} -p #{config.ssh.port} '#{command.join(" ").gsub("'", "'\\\\''")}'".squish
+      "ssh#{ssh_proxy_args} -t #{config.ssh.user}@#{host} -p #{config.ssh.port} '#{command.join(" ").gsub("'", "'\\\\''")}'".squish
```
Comment (npezza93) [inline lib/kamal/commands/base.rb:14]: Hmm i thought i had tests passing with this. But i think with this new approach you can remove squish at the end

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -49,3 +53,5 @@ jobs:
 
       - name: Run tests
         run: bin/test
+        env:
+          RUBYOPT: ${{ matrix.ruby-version == '3.4.0-preview2' && '--enable=frozen-string-literal' || '' }}
```
Comment (Sija) [inline .github/workflows/ci.yml:57]:
```suggestion
          RUBYOPT: ${{ startsWith(matrix.ruby-version, '3.4.') && '--enable=frozen-string-literal' || '' }}
```

File: lib/kamal/commands/base.rb [hunk 9]
```diff
@@ -11,14 +11,7 @@ def initialize(config)
...
     def run_over_ssh(*command, host:)
-      "ssh".tap do |cmd|
-        if config.ssh.proxy && config.ssh.proxy.is_a?(Net::SSH::Proxy::Jump)
-          cmd << " -J #{config.ssh.proxy.jump_proxies}"
-        elsif config.ssh.proxy && config.ssh.proxy.is_a?(Net::SSH::Proxy::Command)
-          cmd << " -o ProxyCommand='#{config.ssh.proxy.command_line_template}'"
-        end
-        cmd << " -t #{config.ssh.user}@#{host} -p #{config.ssh.port} '#{command.join(" ").gsub("'", "'\\\\''")}'"
-      end
+      "ssh#{ssh_proxy_args} -t #{config.ssh.user}@#{host} -p #{config.ssh.port} '#{command.join(" ").gsub("'", "'\\\\''")}'".squish
```
Comment (Sija) [inline lib/kamal/commands/base.rb:14]:
```suggestion
      "ssh #{ssh_proxy_args} -t #{config.ssh.user}@#{host} -p #{config.ssh.port} '#{command.join(" ").gsub("'", "'\\\\''")}'".squish
```

File: lib/kamal/commands/base.rb [hunk 10]
```diff
@@ -92,5 +85,14 @@ def grep(*args)
...
         Kamal::Tags.from_config(config, **details)
       end
+
+      def ssh_proxy_args
+        case config.ssh.proxy
+        when Net::SSH::Proxy::Jump
+          " -J #{config.ssh.proxy.jump_proxies}"
+        when Net::SSH::Proxy::Command
+          " -o ProxyCommand='#{config.ssh.proxy.command_line_template}'"
+        end
```
Comment (Sija) [inline lib/kamal/commands/base.rb:95]:
```suggestion
        when Net::SSH::Proxy::Jump
          "-J #{config.ssh.proxy.jump_proxies}"
        when Net::SSH::Proxy::Command
          "-o ProxyCommand='#{config.ssh.proxy.command_line_template}'"
        end
```

File: lib/kamal/commands/base.rb (hunk 9 above)
Comment (jeremy) [reply lib/kamal/commands/base.rb:14]: Note the spacing. This leaves no double space when there is no proxy so we needn't squish, altering the command.

Comment (jeremy) [inline lib/kamal/commands/base.rb:14]:
```suggestion
      "ssh#{ssh_proxy_args} -t #{config.ssh.user}@#{host} -p #{config.ssh.port} '#{command.join(" ").gsub("'", "'\\\\''")}'"
```

**Review by jeremy — APPROVED** [review]

**Pull request merged**